datafusion/physical-expr/src/expressions/datetime.rs

```diff
@@ -112,71 +120,166 @@ impl PhysicalExpr for DateIntervalExpr {
-                let prior = epoch.add(Duration::days(d as i64));
-                let posterior = do_date_math(prior, scalar, sign)?;
-                let days = posterior.sub(epoch).num_days() as i32;
-                ColumnarValue::Scalar(ScalarValue::Date32(Some(days)))
-            }
-            ScalarValue::Date64(Some(ms)) => {
-                let epoch = NaiveDate::from_ymd(1970, 1, 1);
-                let prior = epoch.add(Duration::milliseconds(ms));
-                let posterior = do_date_math(prior, scalar, sign)?;
-                let ms = posterior.sub(epoch).num_milliseconds();
-                ColumnarValue::Scalar(ScalarValue::Date64(Some(ms)))
-            }
-            ScalarValue::TimestampSecond(Some(ts_s), zone) => {
-                let value = do_data_time_math(ts_s, 0, scalar, sign)?.timestamp();
-                ColumnarValue::Scalar(ScalarValue::TimestampSecond(Some(value), zone))
-            }
-            ScalarValue::TimestampMillisecond(Some(ts_ms), zone) => {
-                let secs = ts_ms / 1000;
-                let nsecs = ((ts_ms % 1000) * 1_000_000) as u32;
-                let value =
-                    do_data_time_math(secs, nsecs, scalar, sign)?.timestamp_millis();
-                ColumnarValue::Scalar(ScalarValue::TimestampMillisecond(
-                    Some(value),
-                    zone,
-                ))
-            }
-            ScalarValue::TimestampMicrosecond(Some(ts_us), zone) => {
-                let secs = ts_us / 1_000_000;
-                let nsecs = ((ts_us % 1_000_000) * 1000) as u32;
-                let value = do_data_time_math(secs, nsecs, scalar, sign)?
-                    .timestamp_nanos()
-                    / 1000;
-                ColumnarValue::Scalar(ScalarValue::TimestampMicrosecond(
-                    Some(value),
-                    zone,
-                ))
-            }
-            ScalarValue::TimestampNanosecond(Some(ts_ns), zone) => {
-                let secs = ts_ns / 1_000_000_000;
-                let nsecs = (ts_ns % 1_000_000_000) as u32;
-                let value =
-                    do_data_time_math(secs, nsecs, scalar, sign)?.timestamp_nanos();
-                ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(value), zone))
-            }
-            _ => Err(DataFusionError::Execution(format!(
-                "Invalid lhs type for DateIntervalExpr: {}",
-                scalar
-            )))?,
-        };
-        Ok(res)
+        match dates {
+            ColumnarValue::Scalar(operand) => evaluate_scalar(operand, sign, intervals),
+            ColumnarValue::Array(array) => evaluate_array(array, sign, intervals),
+        }
     }
 }
 
+pub fn evaluate_array(
+    array: ArrayRef,
+    sign: i32,
+    scalar: &ScalarValue,
+) -> Result<ColumnarValue> {
+    let ret = match array.data_type() {
+        DataType::Date32 => {
+            let array = array.as_any().downcast_ref::<Date32Array>().unwrap();
```

Comment:
I know this shouldn't panic, but since this is returning a result anyway, would it improve auditability to use the error propagation operator here instead?

Reply:
Like this?
```rust
let array = array.as_any().downcast_ref::<Date32Array>().ok_or(
    DataFusionError::Execution("Cast Date32Array error".to_string()),
)?;
```

Reply:
It's a bit more verbose, but I think it makes it easier to tell that it will never panic. @andygrove @alamb do you have thoughts on this?

Reply:
I think by this point in the plan it is ok to panic! as it indicates a bug in the code and I think having to always do the ceremony to check the type makes it less readable. Maybe we could make a function like `as_date32_array` that returned an error (similar to the arrow ones that panic?): https://docs.rs/arrow/20.0.0/arrow/array/index.html#functions ?

Sadly it appears we are inconsistent in the codebase with this regards: https://github.com/apache/arrow-datafusion/search?l=Rust&q=downcast 😢

Reply:
But in other words, I think this PR is ok as written.

Reply:
I will file a ticket as this is confusing and it would be nice to get DataFusion consistent

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/3152